app/src/main/java/io/gnosis/safe/ui/transaction/TransactionsViewModel.kt

```diff
@@ -1,103 +1,122 @@
 import io.gnosis.safe.ui.base.BaseStateViewModel
+import io.gnosis.safe.ui.transaction.list.TransactionPagingRepository
+import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.map
 import pm.gnosis.model.Solidity
 import javax.inject.Inject
 
 class TransactionsViewModel
 @Inject constructor(
-    private val transactionRepository: TransactionRepository,
+    private val transactionsPagingRepository: TransactionPagingRepository,
     private val safeRepository: SafeRepository,
     appDispatchers: AppDispatchers
 ) : BaseStateViewModel<TransactionsViewState>(appDispatchers) {
 
+    private var currentSafeAddress: Solidity.Address? = null
+    private var currentSafeTxItems: Flow<PagingData<TransactionView>>? = null
+
     init {
         safeLaunch {
             safeRepository.activeSafeFlow().collect { load() }
         }
     }
 
-    override fun initialState(): TransactionsViewState = TransactionsViewState(null, true)
+    fun getTransactions(safe: Solidity.Address): Flow<PagingData<TransactionView>> {
 
-    fun load() {
-        safeLaunch {
-            val safeAddress = safeRepository.getActiveSafe()
-            updateState { TransactionsViewState(isLoading = true, viewAction = ActiveSafeChanged(safeAddress)) }
-            if (safeAddress != null) {
-                loadTransactions(safeAddress.address)
-            } else {
-                updateState(forceViewAction = true) { TransactionsViewState(isLoading = false, viewAction = NoSafeSelected) }
-            }
+        val lastResult = currentSafeTxItems
+        if (safe == currentSafeAddress && lastResult != null) {
+            return lastResult
         }
-    }
-
-    private suspend fun loadTransactions(safe: Solidity.Address) {
-        val safeInfo = safeRepository.getSafeInfo(safe)
-        val transactions = transactionRepository.getTransactions(safe, safeInfo)
-        updateState {
-            val mappedTransactions = mapTransactions(transactions, safe)
-            val transactionsWithSectionHeaders = addSectionHeaders(mappedTransactions)
-            TransactionsViewState(
-                isLoading = false,
-                viewAction = mappedTransactions
-                    .takeUnless { it.isEmpty() }
-                    ?.let { LoadTransactions(transactionsWithSectionHeaders) }
-                    ?: ViewAction.ShowEmptyState
-            )
-        }
-    }
-
-    private fun mapTransactions(
-        transactions: Page<Transaction>,
-        safe: Solidity.Address
-    ): List<TransactionView> {
-        return transactions.results.mapNotNull { transaction ->
-            when (transaction) {
-                is Transaction.Transfer -> TransactionView.Transfer(transaction, transaction.recipient == safe)
-                is Transaction.SettingsChange -> TransactionView.SettingsChange(transaction)
-                else -> null
+        currentSafeAddress = safe
```

Comment:
Is this assignment necessary? You can only reach this point if, and only if, `safe` is equal to `currentSafeAddress` anyway.

Reply:
don't see how. Imagine your active safe has changed -> load will be triggered -> current safe still points to the old value -> we don't use cached results, but request new pages and update currentSafeAddress

Reply:
So, in that case using `safeRepository.getActiveSafe()` each time you call this method would be better?